File: pairwiseSwapping.c
//Given a singly linked list of size N. The task is to swap elements in the linked list pair wise. For example, if the input list is 1 2 3 4, the resulting list after swaps will be 2 1 4 3.  

#include <stdio.h>
#include <stdlib.h>

// Node structure
struct Node {
    int data;
    struct Node* next;
};

// Create a new node
struct Node* createNode(int data) {
    struct Node* newnode = (struct Node*)malloc(sizeof(struct Node));
    newnode->data = data;
    newnode->next = NULL;
    return newnode;
}

// Insert node at end
void insertEnd(struct Node** head, int data) {
    struct Node* newnode = createNode(data);
    if (*head == NULL) {
        *head = newnode;
        return;
    }
    struct Node* temp = *head;
    while (temp->next)
        temp = temp->next;
    temp->next = newnode;
}

// Display list
void display(struct Node* head) {
    while (head) {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

// Swap nodes pairwise
void pairwiseSwap(struct Node* head) {
    struct Node* temp = head;

    while (temp != NULL && temp->next != NULL) {
        // Swap data of current node and next node
        int t = temp->data;
        temp->data = temp->next->data;
        temp->next->data = t;

        // Move two nodes ahead
        temp = temp->next->next;
    }
}

// Main
int main() {
    struct Node* head = NULL;
    int n, val,i;

    printf("Enter number of nodes: ");
    scanf("%d", &n);
    printf("Enter %d elements:\n", n);
    for (i = 0; i < n; i++) {
        scanf("%d", &val);
        insertEnd(&head, val);
    }

    printf("\nOriginal List: ");
    display(head);

    pairwiseSwap(head);

    printf("List After Pairwise Swap: ");
    display(head);

    return 0;
}

File: pairwiseSwapping.py
#Given a singly linked list of size N. The task is to swap elements in the linked list pair wise. For example, if the input list is 1 2 3 4, the resulting list after swaps will be 2 1 4 3.
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def insert_end(head, data):
    new = Node(data)
    if not head:
        return new
    temp = head
    while temp.next:
        temp = temp.next
    temp.next = new
    return head

def display(head):
    while head:
        print(head.data, end=" -> ")
        head = head.next
    print("NULL")

def pairwise_swap(head):
    temp = head
    while temp and temp.next:
        temp.data, temp.next.data = temp.next.data, temp.data
        temp = temp.next.next
    return head

# Input and test
n = int(input("Enter number of nodes: "))
print("Enter the elements:")
head = None
for _ in range(n):
    val = int(input())
    head = insert_end(head, val)

print("\nOriginal List:")
display(head)

head = pairwise_swap(head)

print("List After Pairwise Swap:")
display(head)
